spec/user_interface_spec.rb

```diff
@@ -72,4 +72,61 @@
 
     expect(ui.run).to eq(1)
   end
+
+  describe 'ask for fields method' do
+    let(:input) { StringIO.new(contact_inputs(TEST_DETAILS)) }
+    let(:ui) { described_class.new(input, output) }
+
+    it 'asks user for all fields' do
+      ui.ask_for_fields
+
+      expect(output.string).to include(described_class::FIELDS_TO_PROMPTS.values.join("\n"))
+    end
+
+    it 'reads the input for all prompt' do
+      user_input = ui.ask_for_fields
+
+      expect(user_input).to eq(TEST_DETAILS)
+    end
+
+    it 'sends error if number is invalid' do
+      input = StringIO.new(contact_inputs(INVALID_INPUTS))
+      ui = described_class.new(input, output)
+
+      ui.ask_for_fields
+
+      expect(output.string).to include(described_class::ERROR_MESSAGE)
+    end
+
+    it 'sends error if email is invalid' do
+      input = StringIO.new(contact_inputs(INVALID_INPUTS))
+      ui = described_class.new(input, output)
+
+      ui.ask_for_fields
+
+      expect(output.string).to include(described_class::ERROR_MESSAGE)
+    end
+  end
+
+  TEST_DETAILS = {
+    name: 'Matt Damon',
+    address: 'Some address',
+    phone: '08796564231',
+    email: 'matt@damon.com',
+    notes: 'I think he has an Oscar'
+  }.freeze
+
+  INVALID_INPUTS = {
+    name: 'Matt Damon',
+    address: 'Some address',
+    invalid_phone: '08796564',
+    phone: '08796564231',
+    invalid_email: 'mattdamon.com',
+    email: 'matt@damon.com',
+    notes: 'I think he has an Oscar'
+  }.freeze
+
+  def contact_inputs(details)
```

Comment:
You don't need this method :+1:

Reply:
its better to repeat code 3 times instead?

Reply:
now without this method I've had to repeated code 3 times. I think I've miss-understood.

Reply:
In tests you want to be specific and explicit, the rules of production code do not apply to tests, it's ok to be repetitive in tests, as tests are your documentation.

For example, rather than having a method that gets the values of the hash and joins them using `\n`, I would have explicitly typed the values in the test: `"Name\nAddress\nWrong phone\n1234567890\nmatt@damon.com"`.

It's good enough how you have it now though.

Reply:
I dont know were the line is. I feel if I'd put "Name\nAddress\nWrong phone\n1234567890\nmatt@damon.com" in every test you would have call that noise.